File: content/snippets/scala3/interview_lcci_17_12_binode.scala
class TreeNode(var value: Int, var left: TreeNode = null, var right: TreeNode = null)

object Lcci1712BiNode:
  private var prev: TreeNode = _

  private def inorder(root: TreeNode): Unit = {
    if (root == null) return
    inorder(root.left)
    root.left = null
    prev.right = root
    prev = root
    inorder(root.right)
  }

  def convertBiNode(root: TreeNode): TreeNode =
    val dummy = new TreeNode(0)
    prev = dummy
    inorder(root)
    dummy.right

  @main def main(): Unit =
    val root = new TreeNode(4)
    root.left = new TreeNode(2)
    root.right = new TreeNode(5)
    root.left.left = new TreeNode(1)
    root.left.right = new TreeNode(3)
    root.right.right = new TreeNode(6)
    root.left.left.left = new TreeNode(0)

    var head = convertBiNode(root)
    val out = scala.collection.mutable.ArrayBuffer[Int]()
    while (head != null) {
      out += head.value
      head = head.right
    }
    println(out.mkString(" ")) // 0 1 2 3 4 5 6

File: content/snippets/python/interview_lcci_17_12_binode.py
from __future__ import annotations
from typing import Optional


class TreeNode:
    def __init__(self, x: int):
        self.val = x
        self.left: Optional[TreeNode] = None
        self.right: Optional[TreeNode] = None


class Solution:
    def __init__(self) -> None:
        self.prev: Optional[TreeNode] = None

    def convertBiNode(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        dummy = TreeNode(0)
        self.prev = dummy

        def inorder(node: Optional[TreeNode]) -> None:
            if not node:
                return
            inorder(node.left)
            node.left = None
            self.prev.right = node
            self.prev = node
            inorder(node.right)

        inorder(root)
        return dummy.right


def main():
    root = TreeNode(4)
    root.left = TreeNode(2)
    root.right = TreeNode(5)
    root.left.left = TreeNode(1)
    root.left.right = TreeNode(3)
    root.right.right = TreeNode(6)
    root.left.left.left = TreeNode(0)

    head = Solution().convertBiNode(root)
    out = []
    while head:
        out.append(str(head.val))
        head = head.right
    print(' '.join(out))  # 0 1 2 3 4 5 6


if __name__ == "__main__":
    main()
